quickcache/quickcache_helper.py

```diff
@@ -0,0 +1,150 @@
+    def _hash(value, length=32):
+        return hashlib.md5(value).hexdigest()[-length:]
+
+    def _serialize_for_key(self, value):
+        if isinstance(value, basestring):
+            # Unicode and string values should generate the same key since users generally
+            # intend them to mean the same thing. If a use case for differentiating
+            # them presents itself add a 'lenient_strings=False' option to allow
+            # the user to explicitly request the different behaviour.
+            if isinstance(value, unicode):
+                encoded = value.encode('utf-8')
+            else:
+                try:
+                    encoded = value.decode('utf-8').encode('utf-8')
+                except UnicodeDecodeError:
+                    self.encoding_assert(False, 'Non-utf8 encoded string used as cache vary on')
+                    encoded = value
+            return 'u' + self._hash(encoded)
+        elif isinstance(value, bool):
+            return 'b' + str(int(value))
+        elif isinstance(value, (int, long, float)):
+            return 'n' + str(value)
+        elif isinstance(value, (list, tuple)):
+            return 'L' + self._hash(
+                ','.join(map(self._serialize_for_key, value)))
+        elif isinstance(value, dict):
+            return 'D' + self._hash(
+                ','.join(sorted(map(self._serialize_for_key, value.items())))
+            )
+        elif isinstance(value, set):
+            return 'S' + self._hash(
+                ','.join(sorted(map(self._serialize_for_key, value))))
+        elif value is None:
+            return 'N'
+        else:
+            raise ValueError('Bad type "{}": {}'.format(type(value), value))
+
+    def get_cache_key(self, *args, **kwargs):
+        callargs = inspect.getcallargs(self.fn, *args, **kwargs)
+        values = []
+        if isfunction(self.vary_on):
+            values = self.vary_on(*args, **kwargs)
+        else:
+            for arg_name, attrs in self.vary_on:
+                value = callargs[arg_name]
+                for attr in attrs:
+                    value = getattr(value, attr)
+                values.append(value)
+        args_string = ','.join(self._serialize_for_key(value)
+                               for value in values)
+        if len(args_string) > 150:
+            args_string = 'H' + self._hash(args_string)
+        return 'quickcache.{}/{}'.format(self.prefix, args_string)
+
+    def skip(self, *args, **kwargs):
+        if not self.skip_arg:
+            return False
+        elif isinstance(self.skip_arg, basestring):
+            callargs = inspect.getcallargs(self.fn, *args, **kwargs)
+            return callargs[self.skip_arg]
+        elif isfunction(self.skip_arg):
+            return self.skip_arg(*args, **kwargs)
+        else:
+            assert False, "skip_arg must be None, a string, or a function " \
+                          "and this should have been checked in __init__"
```

Comment:
This `skip` method is adapted from the stuff at the beginning of `SkippableQuickCache` (RIP)'s overridden `__call__` method.